# 4-垃圾邮件检测

现在，我们将进行第二部分的练习。 在这一部分中，我们的目标是使用SVM来构建垃圾邮件过滤器。 在练习文本中，有一个任务涉及一些文本预处理，以获得适合SVM处理的格式的数据。 然而，这个任务很简单（将字词映射到为练习提供的字典中的ID），而其余的预处理步骤（如HTML删除，词干，标准化等）已经完成。 我将跳过机器学习任务，而不是重现这些预处理步骤，其中包括从预处理过的训练集构建分类器，以及将垃圾邮件和非垃圾邮件转换为单词出现次数的向量的测试数据集。

In [1]:
from sklearn import svm, metrics
from sklearn.linear_model import LogisticRegression

import scipy.io as sio

D:\Programming tools\Python\Python36\lib\importlib\_bootstrap.py:205: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)


In [4]:
spam_train = sio.loadmat('data/spamTrain.mat')
spam_test = sio.loadmat('data/spamTest.mat')
spam_train.keys(),spam_test.keys()

(dict_keys(['__header__', '__version__', '__globals__', 'X', 'y']),
 dict_keys(['__header__', '__version__', '__globals__', 'Xtest', 'ytest']))

In [7]:
X, y = spam_train['X'], spam_train['y'].ravel()
Xtest, ytest = spam_test['Xtest'], spam_test['ytest'].ravel()

X.shape, y.shape, Xtest.shape, ytest.shape

((4000, 1899), (4000,), (1000, 1899), (1000,))

每个文档已经转换为一个向量， 其中1899个维对应于词汇表中的1899个单词。它们的值为二进制，表示文档中是否存在单词。在这一点上，训练评估是用一个分类器拟合测试数据的问题。

In [8]:
svc = svm.SVC()
svc.fit(X,y)

SVC()

In [10]:
pred = svc.predict(Xtest)
print(metrics.classification_report(ytest, pred))


              precision    recall  f1-score   support

           0       0.99      1.00      0.99       692
           1       0.99      0.97      0.98       308

    accuracy                           0.99      1000
   macro avg       0.99      0.98      0.98      1000
weighted avg       0.99      0.99      0.99      1000



使用logistic regresion试试

In [13]:
logit = LogisticRegression()
logit.fit(X,y)

LogisticRegression()

In [14]:
logit.score(X,y)

0.99975

In [15]:
logit.score(Xtest,ytest)

0.994

In [12]:
pred = logit.predict(Xtest)
print(metrics.classification_report(ytest, pred))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00       692
           1       0.99      0.99      0.99       308

    accuracy                           0.99      1000
   macro avg       0.99      0.99      0.99      1000
weighted avg       0.99      0.99      0.99      1000

